image/image_test.go

```diff
@@ -0,0 +1,301 @@
+        "AttachStdin": false,
+        "AttachStdout": false,
+        "Cmd": [
+            "sh"
+        ],
+        "Domainname": "",
+        "Entrypoint": null,
+        "Env": [
+            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
+        ],
+        "Hostname": "55cd1f8f6e5b",
+        "Image": "sha256:e732471cb81a564575aad46b9510161c5945deaf18e9be3db344333d72f0b4b2",
+        "Labels": {},
+        "OnBuild": null,
+        "OpenStdin": false,
+        "StdinOnce": false,
+        "Tty": false,
+        "User": "",
+        "Volumes": null,
+        "WorkingDir": ""
+    },
+    "container": "764ef4448baa9a1ce19e4ae95f8cdd4eda7a1186c512773e56dc634dff208a59",
+    "container_config": {
+        "AttachStderr": false,
+        "AttachStdin": false,
+        "AttachStdout": false,
+        "Cmd": [
+            "/bin/sh",
+            "-c",
+            "#(nop) CMD [\"sh\"]"
+        ],
+        "Domainname": "",
+        "Entrypoint": null,
+        "Env": [
+            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
+        ],
+        "Hostname": "55cd1f8f6e5b",
+        "Image": "sha256:e732471cb81a564575aad46b9510161c5945deaf18e9be3db344333d72f0b4b2",
+        "Labels": {},
+        "OnBuild": null,
+        "OpenStdin": false,
+        "StdinOnce": false,
+        "Tty": false,
+        "User": "",
+        "Volumes": null,
+        "WorkingDir": ""
+    },
+    "created": "2016-06-23T23:23:37.198943461Z",
+    "docker_version": "1.10.3",
+    "history": [
+        {
+            "created": "2016-06-23T23:23:36.73131105Z",
+            "created_by": "/bin/sh -c #(nop) ADD file:9ca60502d646bdd815bb51e612c458e2d447b597b95cf435f9673f0966d41c1a in /"
+        },
+        {
+            "created": "2016-06-23T23:23:37.198943461Z",
+            "created_by": "/bin/sh -c #(nop) CMD [\"sh\"]",
+            "empty_layer": true
+        }
+    ],
+    "os": "linux",
+    "rootfs": {
+        "diff_ids": [
+            "sha256:8ac8bfaff55af948c796026ee867448c5b5b5d9dd3549f4006d9759b25d4a893"
+        ],
+        "type": "layers"
+    }
+}
+`
+)
+
+var (
+	refStr = `{"digest":"<manifest_digest>","mediaType":"application/vnd.oci.image.manifest.v1+json","size":<manifest_size>}`
+
+	manifestStr = `{
+    "annotations": null,
+    "config": {
+        "digest": "<config_digest>",
+        "mediaType": "application/vnd.oci.image.config.v1+json",
+        "size": <config_size>
+    },
+    "layers": [
+        {
+            "digest": "<layer_digest>",
+            "mediaType": "application/vnd.oci.image.layer.tar+gzip",
+            "size": <layer_size>
+        }
+    ],
+    "mediaType": "application/vnd.oci.image.manifest.v1+json",
+    "schemaVersion": 2
+}
+ `
+)
+
+func TestValidateLayout(t *testing.T) {
+	root, err := ioutil.TempDir("", "oci-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(root)
+
+	err = os.MkdirAll(filepath.Join(root, "blobs", "sha256"), 0700)
```

Comment:
On Wed, Sep 14, 2016 at 12:43:24AM -0700, xiekeyang wrote:
> > +	err = os.MkdirAll(filepath.Join(root, "blobs", "sha256"), 0700)
>
> > And “create image-layout fodder for testing” is likely to be
> > useful stuff in lots of places. Can we make layer-generator,
> > etc. helpers instead of baking it into
> > TestUnpackLayerRemovePartialyUnpackedFile. I'm missing #159 and
> > 87d6a53 ;).
>
> @wking Got it. Can we place this code to a function `createTarFile`
> in `*_test.go`? For now this is only used by unit test, and had
> better not to be compiled to binary.  For that I will fix
> `image_test.go` and `manifest_test.go` to reuse the `func
> createTarFile`

This sounds fine to me.  I'd encourage helpers for:

* create image-layout (directory or tar file) [1]
* create dummy layer
* PutJSON [2]
* PutRef [3]

[1]: https://github.com/opencontainers/image-spec/pull/159/commits/87d6a5351fdcc738b60f5b1612ff528307f8dbe8#diff-745d1b74657ed8e87ac8c4e3f520f73aR223
[2]: https://github.com/opencontainers/image-spec/pull/310#discussion_r78697964
[3]: https://github.com/opencontainers/image-spec/pull/159/commits/2d555835b84d5cf7deaa9b108efac43e683a372e#diff-d8af84c58f6ae3369e0b5cbca06727dfR30